File: 1863_SumOfAllSubsetXORTotals.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:

    //both codes are correct

    int subsetXORSum(vector<int>& nums) {
        int ans = 0;
        for(int n: nums){
            ans|=n;
        }
        return ans<<(nums.size()-1);
    }


    // int subsetXORSum(vector<int>& nums) {
    //     return getXOR(nums, nums.size(), 0, 0);
    // }

    // int getXOR(vector<int>& nums, int n, int i, int running_ans) {
    //     if (i==n){
    //         return running_ans; 
    //     }
    //     return getXOR(nums,n,i+1, running_ans^nums[i]) + getXOR(nums,n,i+1, running_ans);

    // }

};

File: 1863_SumOfAllSubsetXORTotals.py
from typing import List

# TC = O(N), SC = O(1)
class Solution:
    def subsetXORSum(self, nums: List[int]) -> int:
        setBits = 0
        for num in nums:
            setBits |= num
        
        return setBits<<(len(nums)-1)
    

# TC = O(N * 2**N), SC = O(n)
# class Solution:
#     def subsetXORSum(self, nums: List[int]) -> int:
#         n = len(nums)

#         def traverse(i: int, currentXor:int ):
#             if i==n:
#                 return currentXor
            
#             return traverse(i+1, currentXor) + traverse(i+1, currentXor^nums[i])

#         return traverse(0,0)